miner/proposal_builder.go

```diff
@@ -374,137 +335,370 @@ func (pb *ProposalBuilder) decideMeshHash(ctx context.Context, current types.Lay
+		set = append(set, ownAtx)
+		totalWeight += ownWeight
 	}
+	return totalWeight, set, nil
+}
 
-	pb.eg.Go(func() error {
-		// generate a new requestID for the new proposal message
-		newCtx := log.WithNewRequestID(ctx, layerID, p.ID())
-		// validation handler, where proposal is persisted, is applied synchronously before
-		// proposal is sent over the network
-		data, err := codec.Encode(p)
+func generateActiveSet(
+	logger log.Log,
+	cdb *datastore.CachedDB,
+	signer *signing.VRFSigner,
+	target types.EpochID,
+	epochStart time.Time,
+	goodAtxPercent int,
+	networkDelay time.Duration,
+	ownAtx types.ATXID,
+	ownWeight uint64,
+) (uint64, []types.ATXID, error) {
+	var (
+		totalWeight uint64
+		set         []types.ATXID
+		numOmitted  = 0
+	)
+	if err := cdb.IterateEpochATXHeaders(target, func(header *types.ActivationTxHeader) error {
+		grade, err := gradeAtx(cdb, header.NodeID, header.Received, epochStart, networkDelay)
 		if err != nil {
-			pb.logger.WithContext(newCtx).With().Fatal("failed to serialize proposal", log.Err(err))
+			return err
 		}
-		if err = pb.publisher.Publish(newCtx, pubsub.ProposalProtocol, data); err != nil {
-			pb.logger.WithContext(newCtx).With().Error("failed to send proposal", log.Err(err))
-		} else {
-			events.EmitProposal(layerID, p.ID())
-			events.ReportProposal(events.ProposalCreated, p)
+		if grade != good && header.NodeID != signer.NodeID() {
+			logger.With().Info("atx omitted from active set",
+				header.ID,
+				log.Int("grade", int(grade)),
+				log.Stringer("smesher", header.NodeID),
+				log.Bool("own", header.NodeID == signer.NodeID()),
+				log.Time("received", header.Received),
+				log.Time("epoch_start", epochStart),
+			)
+			numOmitted++
+			return nil
 		}
+		totalWeight += header.GetWeight()
+		set = append(set, header.ID)
 		return nil
+	}); err != nil {
+		return 0, nil, err
+	}
+
+	if total := numOmitted + len(set); total == 0 {
+		return 0, nil, fmt.Errorf("empty active set")
+	} else if numOmitted*100/total > 100-goodAtxPercent {
+		// if the node is not synced during `targetEpoch-1`, it doesn't have the correct receipt timestamp
+		// for all the atx and malfeasance proof. this active set is not usable.
+		// TODO: change after timing info of ATXs and malfeasance proofs is sync'ed from peers as well
+		var err error
+		totalWeight, set, err = activesFromFirstBlock(cdb, signer, target, ownAtx, ownWeight)
+		if err != nil {
+			return 0, nil, err
+		}
+		logger.With().Info("miner not synced during prior epoch, active set from first block",
+			log.Int("all atx", total),
+			log.Int("num omitted", numOmitted),
+			log.Int("num block atx", len(set)),
+		)
+	} else {
+		logger.With().Info("active set selected for proposal using grades",
+			log.Int("num atx", len(set)),
+			log.Int("num omitted", numOmitted),
+			log.Int("min atx good pct", goodAtxPercent),
+		)
+	}
+	sort.Slice(set, func(i, j int) bool {
+		return bytes.Compare(set[i].Bytes(), set[j].Bytes()) < 0
 	})
```

Comment:
Why does it need to be sorted?

Reply:
activeset later hashed and referenced as in ballot/proposal . we want to keep number of unique activesets low, as otherwise they bloat network and state